soweego/validator/checks.py

```diff
@@ -629,84 +784,138 @@ def _validate(criterion, wd, target_generator, to_be_deprecated, to_be_added):
+    # Biographical validation requires more complex comparisons
+    elif criterion == keys.BIODATA:
+        # `wd_data` has either couples or triples: couples are dates
+        wd_dates = set(filter(lambda x: len(x) == 2, wd_data))
+        # No cast to `set` because `wd_data` triples hold sets themselves
+        wd_other = list(filter(lambda x: len(x) == 3, wd_data))
+        # In `target_data` we look for relevant date PIDs
+        target_dates = set(
+            filter(
+                lambda x: x[0]
+                in (vocabulary.DATE_OF_BIRTH, vocabulary.DATE_OF_DEATH),
+                target_data,
+            )
+        )
+        target_other = target_data.difference(target_dates)
+        shared_dates, extra_dates = _compare('dates', wd_dates, target_dates)
+        shared_other, extra_other = _compare('other', wd_other, target_other)
+        shared = shared_dates | shared_other
+        extra = extra_dates | extra_other
+    else:
+        raise ValueError(
+            f"Invalid validation criterion: '{criterion}'. "
+            f"Please use either '{keys.LINKS}' or '{keys.BIODATA}'"
+        )
 
     return shared, extra
 
 
-def _extract_dates(data):
-    dates = set()
-    for pid, value in data:
-        if pid in (vocabulary.DATE_OF_BIRTH, vocabulary.DATE_OF_DEATH):
-            dates.add((pid, value))
-    # Remove dates from input set
-    data.difference_update(dates)
-    return dates
-
-
-def _compare_dates(wd, target):
-    shared_dates, extra_dates = set(), set()
-
-    for wd_elem, t_elem in zip_longest(wd, target):
-        # Skip pair with None elements
-        if None in (wd_elem, t_elem):
-            continue
-
-        wd_pid, wd_val = wd_elem
-        t_pid, t_val = t_elem
-
-        # Don't compare birth with death dates
-        if wd_pid != t_pid:
-            continue
-
-        # Skip unexpected None values
-        if None in (wd_val, t_val):
-            LOGGER.warning(
-                'Skipping unexpected %s date pair with missing value(s)',
-                (wd_elem, t_elem),
+def _compare(what, wd, target):
+    shared, extra = set(), set()
+    # Keep track of matches to avoid useless computation
+    # and incorrect comparisons:
+    # this happens when WD has multiple claims with
+    # the same property
+    wd_matches, target_matches = [], []
+
+    for i, wd_elem in enumerate(wd):
+        for j, t_elem in enumerate(target):
+            # Don't compare when already matched
+            if i in wd_matches or j in target_matches:
+                continue
+
+            # Don't compare different PIDs
+            if wd_elem[0] != t_elem[0]:
+                continue
+
+            # Skip unexpected `None` values
+            if None in (wd_elem[1], t_elem[1]):
+                LOGGER.warning(
+                    'Skipping unexpected %s pair with missing value(s)',
+                    (wd_elem, t_elem),
+                )
+                continue
+
+            inputs = (
```

Comment:
at this point why don't you put commas on their own line ?! (jk, but are you using black? I hate black for its utter disregard and waste of vertical space - just personal preference no need to change anything)